cmd/kubernetes-discovery/discoverysummarizer/discoverysummarizer.go

```diff
@@ -0,0 +1,186 @@
+	}
+	w.WriteHeader(statusCode)
+	w.Write([]byte("Success"))
+}
+
+// Handler for group versions summarizer.
+func (ds *discoverySummarizerServer) summarizeGroupVersionsHandler(path string) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var apiGroupList *unversioned.APIGroupList
+		for _, serverAddress := range ds.groupVersionPaths[path] {
+			groupList, err := ds.getAPIGroupList(serverAddress + path)
+			if err != nil {
+				ds.writeErr(http.StatusBadGateway, err, w)
+				return
+			}
+			if apiGroupList == nil {
+				apiGroupList = &unversioned.APIGroupList{}
+				*apiGroupList = *groupList
+			} else {
+				apiGroupList.Groups = append(apiGroupList.Groups, groupList.Groups...)
+			}
+		}
+		ds.writeRawJSON(http.StatusOK, *apiGroupList, w)
+		return
+	}
+}
+
+// Handler for legacy versions summarizer.
+func (ds *discoverySummarizerServer) summarizeLegacyVersionsHandler(path string) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		if len(ds.legacyVersionPaths[path]) > 1 {
+			err := fmt.Errorf("invalid multiple servers serving legacy group %v", ds.legacyVersionPaths[path])
+			ds.writeErr(http.StatusInternalServerError, err, w)
+		}
+		serverAddress := ds.legacyVersionPaths[path][0]
+		apiVersions, err := ds.getAPIVersions(serverAddress + path)
+		if err != nil {
+			ds.writeErr(http.StatusBadGateway, err, w)
+			return
+		}
+		ds.writeRawJSON(http.StatusOK, apiVersions, w)
+		return
+	}
+}
+
+func (ds *discoverySummarizerServer) getAPIGroupList(serverAddress string) (*unversioned.APIGroupList, error) {
+	response, err := http.Get(serverAddress)
+	if err != nil {
+		return nil, fmt.Errorf("Error in fetching %s: %v", serverAddress, err)
+	}
+	defer response.Body.Close()
+	contents, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Error reading response from %s: %v", serverAddress, err)
+	}
+	var apiGroupList unversioned.APIGroupList
+	err = json.Unmarshal(contents, &apiGroupList)
+	if err != nil {
+		return nil, fmt.Errorf("Error in unmarshalling response from server %s: %v", serverAddress, err)
+	}
+	return &apiGroupList, nil
+}
+
+func (ds *discoverySummarizerServer) getAPIVersions(serverAddress string) (*unversioned.APIVersions, error) {
+	response, err := http.Get(serverAddress)
+	if err != nil {
+		return nil, fmt.Errorf("Error in fetching %s: %v", serverAddress, err)
+	}
+	defer response.Body.Close()
+	contents, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, fmt.Errorf("Error reading response from %s: %v", serverAddress, err)
+	}
+	var apiVersions unversioned.APIVersions
+	err = json.Unmarshal(contents, &apiVersions)
+	if err != nil {
+		return nil, fmt.Errorf("Error in unmarshalling response from server %s: %v", serverAddress, err)
+	}
+	return &apiVersions, nil
+}
+
+func (ds *discoverySummarizerServer) writeRawJSON(statusCode int, object interface{}, w http.ResponseWriter) {
```

Comment:
I would feel a lot better if we actually used runtime.Object instead of interface{}, and if we used the real encoding/decoding stack. However, if you add a TODO I can live without for the first commit.